### DISCUSSION TITLE
Configuration Context and Templates - Host Var defined via DataSource

### DISCUSSION THREAD

**Discussion by Kani999** [netbox-community/netbox#12740] [General]
Hello,
I'm playing with Config Context and Data Sources.
I would like to pull "host variables" from my git repository and store it to the device. But the device itself has only Local Context Data which can be added only via JSON in the GUI.
Will it be replaced or modified to add host variables via a data source definition in some future release?

Also, I would like to append multiple Render Config Templates to the device.
So I could for example generate "running config" like template and a second template for the NetConf like running-config (XML)
Or just create smaller chunks of configuration.

For demonstration. My config template for device looks like this.
```
! Router configuration
{% include 'templates/common/hostname.j2' %}
{% include 'templates/common/intf.j2' %}
{% include 'templates/common/ntp.j2' %}
{% include 'templates/iosxe/syslog.j2' %}
```
This template includes parts of other templates.  But I'm not able to render just the Interfaces config. 

---
For the 1st problem (assigning the device context data) I'm using Custom Field (local_context_data) which points to a Object of DataSource. 
Then I can render the part of the template like: 
```
{% if device.cf['local_config_context']%}
{% for vrf in device.cf['local_config_context'].data["vrfs"] %}
vrf {{ vrf.name }
.
.
>
{% endif %}
```

Which refers to a `device.cf['local_config_context']` the caveat of this is, that the configuration data are not visible in https://netbox-instance/dcim/devices/DEVICE_ID/config-context/.

---

For the 2nd problem (multiple config templates for a device) I'm trying to use Custom Field which refers to a Multiple Object of Configuration Template. 
In this way, I think it is possible to render parts of template or a different templates at all but only via a CMD using the API calls. 

---

Did somebody try some different ways to do these things?

**Comment by candlerb** [comment]
> Also, I would like to append multiple Render Config Templates to the device.

Right now, it's only a single config template per device. But what you could do as a workaround is to generate a single file with delimited sections, like the [go playground](https://go.dev/play/p/BWJ4dcUqVfT) does.

```
-- running-config --
! Config goes here
end
-- netconf.xml --
<foo>
</foo>
```

(I think there is a name for this format, but I can't find it right now)

**Reply by Kani999** [reply]
Yeah, I thought about a JSON format, where I could pick the part of config I want, but when everything is print out int the UI is little bit unreadable.

**Reply by candlerb** [reply]
YAML with multi-line strings might work:

```
---
config: |
  ! Config goes here
  end
netconf: |
  <foo>
  </foo>
```

Or even multi-document [scalars](https://yaml.org/spec/1.2.2/#23-scalars):

```
--- |
! Config goes here
end

--- |
<foo>
</foo>
```

Both are human readable but reasonably standard to machine-parse.
